<!-- TOC -->

- [最初的最初](#最初的最初)
- [中世界JavaScript的异步（Promise）](#中世界javascript的异步promise)
- [异步结合 async/await使用](#异步结合-asyncawait使用)
- [EventEmitter模块](#eventemitter模块)
- [Events !== Asynchrony](#events--asynchrony)
- [异步事件](#异步事件)

<!-- /TOC -->

# 最初的最初
实现`EventEmitter`接口，利用回调。但是回调并不意味这异步调用，调用回调可以是同步，也可以是异步。比如以下代码（一个设计不好的函数，因为从设计风格上说，传入的回调要不全是同步调用，要不全是异步调用，混用会导致不一致）：
```js
function fileSize (fileName, cb) {
  if (typeof fileName !== 'string') {
    return cb(new TypeError('argument should be string')); // 同步
  }
  fs.stat(fileName, (err, stats) => {
    if (err) { return cb(err); } // 异步
    cb(null, stats.size); // 异步
  });
}
```
异步调用回调的例子：
```js
const readFileAsArray = function(file, cb) {
  fs.readFile(file, function(err, data) {
    if (err) {
      return cb(err);
    }
    const lines = data.toString().trim().split('\n');
    cb(null, lines);
  });
};
```
这样使用：
```js
readFileAsArray('./numbers.txt', (err, lines) => {
  if (err) throw err;
  const numbers = lines.map(Number);
  const oddNumbers = numbers.filter(n => n%2 === 1);
  console.log('Odd numbers count:', oddNumbers.length);
});
```

一般来说，回调函数是`error first`的风格。也就是第一个参数是错误码，第二个参数是负载。

# 中世界JavaScript的异步（Promise）
`Promise`可以看作是回调的替代。`Promise`对象一般处理`成功`和`失败`场景，同时允许链式调用。

上述同样的`readFileAsArray`，如果是`Promise`风格：
```js
readFileAsArray('./numbers.txt')
  .then(lines => {
    const numbers = lines.map(Number);
    const oddNumbers = numbers.filter(n => n%2 === 1);
    console.log('Odd numbers count:', oddNumbers.length);
  })
  .catch(console.error);
```

那么对应的，就应该对`readFileAsArray`做相应的修改：
```js
const readFileAsArray = function(file, cb = () => {}) {
  return new Promise((resolve, reject) => {
    fs.readFile(file, function(err, data) {
      if (err) {
        reject(err);
        return cb(err);
      }
      const lines = data.toString().trim().split('\n');
      resolve(lines);
      cb(null, lines);
    });
  });
};
```
`cb`有个默认值。

# 异步结合 async/await使用
如果说`Promise`提升了我们写异步代码的便捷性的话，那么使用`async/await`则更大程度上提升了代码的可阅读性。上述的例子就可以这么写：
```js
async function countOdd () {
  try {
    const lines = await readFileAsArray('./numbers');
    const numbers = lines.map(Number);
    const oddCount = numbers.filter(n => n%2 === 1).length;
    console.log('Odd numbers count:', oddCount);
  } catch(err) {
    console.error(err);
  }
}
countOdd();
```
一般来说，`async`函数的调用外层都会包装一层`try / catch`，用来处理`error`情况。对于支持`Promise`借口的函数，我们都可以按照上述这么使用。但是不能用于`回调方式`的异步函数。

# EventEmitter模块
`EventEmitter`模块是`Node`的异步事件驱动架构的核心。它主要有两个特性：
- 触发命名的事件，通过`emit`
- 订阅和反订阅事件监听函数，通过`on`和`removeListener`

# Events !== Asynchrony
比如有一个`Logger`类：
```js
const EventEmitter = require('events');

class WithLog extends EventEmitter {
  execute(taskFunc) {
    console.log('Before executing');
    this.emit('begin');
    taskFunc();
    this.emit('end');
    console.log('After executing');
  }
}

const withLog = new WithLog();

withLog.on('begin', () => console.log('About to execute'));
withLog.on('end', () => console.log('Done with execute'));

withLog.execute(() => console.log('*** Executing task ***'));
```

输出：
```js
Before executing
About to execute
*** Executing task ***
Done with execute
After executing
```

需要 __注意非常重要__ 的一点是，上面的代码都是同步执行的，因为`execute`的回调是同步回调。但是如果传入异步的回调，则会得到不一致的结果。
```js
// ...

withLog.execute(() => {
  setImmediate(() => {
    console.log('*** Executing task ***')
  });
});
```

输出：
```js
Before executing
About to execute
Done with execute
After executing
*** Executing task ***
```

意思是说，在 __回调（or Promise）__ 和 __基于事件通信__ 之间存在一个鸿沟。相比于`回调`，`基于事件通信`的好处是对于同一个信号，可以有多个监听函数。

# 异步事件
```js
const fs = require('fs');
const EventEmitter = require('events');

class WithTime extends EventEmitter {
  execute(asyncFunc, ...args) {
    this.emit('begin');
    console.time('execute');
    asyncFunc(...args, (err, data) => {
      if (err) {
        return this.emit('error', err);
      }

      this.emit('data', data);
      console.timeEnd('execute');
      this.emit('end');
    });
  }
}

const withTime = new WithTime();

withTime.on('begin', () => console.log('About to execute'));
withTime.on('end', () => console.log('Done with execute'));

withTime.execute(fs.readFile, __filename);
```

上述例子就将原先通过回调获取到的数据，现在可以通过订阅事件获取。同时也可以将上述代码重构成`async\await`风格：
```js
async execute(asyncFunc, ...args) {
    this.emit('begin');
    try {
      console.time('execute');
      const data = await asyncFunc(...args);
      this.emit('data', data);
      console.timeEnd('execute');
      this.emit('end');
    } catch(err) {
      this.emit('error', err);
    }
  }
```

但是对于上面的测试例子，`fs.readFile`是不支持返回`Propmise`对象的。对于使用`on`订阅的监听，执行监听函数的顺序是先订阅先执行。可以使用`prependListener`插到监听队列的前面。